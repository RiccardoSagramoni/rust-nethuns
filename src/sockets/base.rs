use std::ffi::CString;
use std::fmt::{self, Debug, Display};
use std::marker::PhantomData;
use std::mem;
use std::sync::atomic;

use derivative::Derivative;
use getset::{CopyGetters, Getters, Setters};

use crate::types::{NethunsFilter, NethunsQueue, NethunsSocketOptions};

// TODO
// use super::pcap::NethunsSocketPcap;
use super::api::Pkthdr;
use super::ring::{AtomicRingSlotStatus, NethunsRing, RingSlotStatus};
use super::PkthdrTrait;


/// Base structure for a `NethunsSocket`.
///
/// This data structure is common to all the implementation of a "nethuns socket",
/// for the supported underlying I/O frameworks. Thus, it's independent from
/// low-level implementation of the sockets.
#[derive(Default, Derivative, Getters, Setters, CopyGetters)]
#[derivative(Debug)]
#[getset(get = "pub")]
pub struct NethunsSocketBase {
    /// Configuration options
    pub(super) opt: NethunsSocketOptions,
    
    /// Rings used for transmission
    pub(super) tx_ring: Option<NethunsRing>,
    
    /// Rings used for reception
    pub(super) rx_ring: Option<NethunsRing>,
    
    /// Name of the binded device
    pub(super) devname: CString,
    
    /// Queue binded to the socket
    #[getset(get_copy = "pub with_prefix")]
    pub(super) queue: NethunsQueue,
    
    /// Index of the interface
    #[getset(get_copy = "pub with_prefix")]
    pub(super) ifindex: libc::c_int,
    
    /// Closure used for filtering received packets.
    #[derivative(Debug = "ignore")]
    #[getset(set = "pub")]
    pub(super) filter: Option<Box<NethunsFilter>>,
    // errbuf removed => use Result as return type
    // filter_ctx removed => use closures with move semantics
}


//


/// Packet received when calling [`NethunsSocket::recv()`](crate::sockets::NethunsSocket::recv)
/// or [`NethunsSocketPcap::read()`](crate::sockets::pcap::NethunsSocketPcap::read).
///
/// The struct contains a [`PhantomData`] marker associated with the socket itself,
/// so that the `RecvPacket` item is valid as long as the socket is alive.
#[derive(Debug)]
pub struct RecvPacket<'a, T> {
    data: RecvPacketData<'a>,
    
    phantom_data: PhantomData<&'a T>,
}


impl<'a, T> RecvPacket<'a, T> {
    pub(super) fn new<'b>(
        data: RecvPacketData<'b>,
        phantom_data: PhantomData<&'a T>,
    ) -> Self {
        let data: RecvPacketData<'a> = unsafe {
            // [SAFETY] As long as the socket is alive, the references
            // to the data are valid
            RecvPacketData {
                id: data.id,
                pkthdr: mem::transmute(data.pkthdr),
                buffer: mem::transmute(data.buffer),
                slot_status_flag: mem::transmute(data.slot_status_flag),
            }
        };
        RecvPacket { data, phantom_data }
    }
    
    #[inline(always)]
    pub fn id(&self) -> usize {
        self.data.id
    }
    
    #[inline(always)]
    pub fn pkthdr(&self) -> &dyn PkthdrTrait {
        self.data.pkthdr
    }
    
    #[inline(always)]
    pub fn buffer(&self) -> &[u8] {
        self.data.buffer
    }
}


impl<T> Display for RecvPacket<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{{\n    id: {},\n    pkthdr: {:?},\n    buffer: {:?}\n}}",
            self.id(),
            self.pkthdr(),
            self.buffer()
        )
    }
}


//


/// Packet received when calling [`NethunsSocket::recv()`](crate::sockets::NethunsSocket::recv)
/// or [`NethunsSocketPcap::read()`](crate::sockets::pcap::NethunsSocketPcap::read)
/// with static lifetime.
///
/// It **must** be wrapped inside `RecvPacket` struct before being handed to the user.
#[derive(Debug)]
pub(super) struct RecvPacketData<'a> {
    id: usize,
    pkthdr: &'a dyn PkthdrTrait,
    
    buffer: &'a [u8],
    
    /// Reference used to set the status flag of the corresponding ring slot
    /// to `Free` when the `RecPacketData` is dropped.
    slot_status_flag: &'a AtomicRingSlotStatus,
}

impl<'a> RecvPacketData<'a> {
    pub fn new(
        id: usize,
        pkthdr: &'a Pkthdr,
        buffer: &'a [u8],
        slot_status_flag: &'a AtomicRingSlotStatus,
    ) -> Self {
        Self {
            id,
            pkthdr,
            buffer,
            slot_status_flag,
        }
    }
}

impl Drop for RecvPacketData<'_> {
    /// Release the buffer by resetting the status flag of
    /// the corresponding ring slot.
    fn drop(&mut self) {
        self.slot_status_flag
            .store(RingSlotStatus::Free, atomic::Ordering::Release);
    }
}
