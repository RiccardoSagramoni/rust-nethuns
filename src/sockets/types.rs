// TODO

use derive_builder::Builder;
use derive_new::new;


#[derive(Clone, Builder, Debug, Default, new, PartialEq, PartialOrd)]
#[builder(pattern = "owned", default)]
pub struct NethunsPkthdrType {
	
}
