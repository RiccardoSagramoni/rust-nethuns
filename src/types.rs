use derive_builder::Builder;
use getset::Getters;


#[derive(Clone, Copy, Debug, Default, PartialEq, PartialOrd)]
pub enum NethunsQueue {
    Some(u32),
    #[default]
    Any,
}


#[derive(Clone, Copy, Debug, Default, PartialEq, PartialOrd)]
pub enum NethunsCaptureDir {
    #[default]
    In,
    Out,
    InOut,
}


#[derive(Clone, Copy, Debug, Default, PartialEq, PartialOrd)]
pub enum NethunsCaptureMode {
    #[default]
    Default,
    SkbMode,
    DrvMode,
    ZeroCopy,
}


#[derive(Clone, Copy, Debug, Default, PartialEq, PartialOrd)]
pub enum NethunsSocketMode {
    #[default]
    RxTx,
    RxOnly,
    TxOnly,
}


/// Options for the nethuns socket.
#[derive(Builder, Clone, Debug, Default, PartialEq, PartialOrd)]
#[builder(pattern = "owned", default)]
pub struct NethunsSocketOptions {
    pub numblocks: u32,
    pub numpackets: u32,
    pub packetsize: u32,
    pub timeout_ms: u32,
    pub dir: NethunsCaptureDir,
    pub capture: NethunsCaptureMode,
    pub mode: NethunsSocketMode,
    pub promisc: bool,
    pub rxhash: bool,
    pub tx_qdisc_bypass: bool,
    
    /// xdp only
    pub xdp_prog: Option<String>,
    /// xdp only   
    pub xdp_prog_sec: Option<String>,
    /// xdp only
    pub xsk_map_name: Option<String>,
    /// xdp only
    pub reuse_maps: Option<bool>,
    /// xdp only
    pub pin_dir: Option<String>,
}


/// Statistics for the nethuns socket.
#[derive(Clone, Copy, Debug, Default, PartialEq, PartialOrd, Getters)]
#[getset(get = "pub")]
pub struct NethunsStat {
    rx_packets: u64,
    tx_packets: u64,
    rx_dropped: u64,
    rx_if_dropped: u64,
    /// xdp only
    rx_invalid: u64,
    /// xdp only
    tx_invalid: u64,
    freeze: u64,
}


#[cfg(test)]
mod tests {
    use is_trait::is_trait;
    
    
    #[test]
    fn assert_send_trait() {
        assert!(is_trait!(super::NethunsCaptureDir, Send));
        assert!(is_trait!(super::NethunsCaptureMode, Send));
        assert!(is_trait!(super::NethunsSocketMode, Send));
        assert!(is_trait!(super::NethunsSocketOptions, Send));
        assert!(is_trait!(super::NethunsStat, Send));
    }
    
    
    #[test]
    fn assert_sync_trait() {
        assert!(is_trait!(super::NethunsCaptureDir, Sync));
        assert!(is_trait!(super::NethunsCaptureMode, Sync));
        assert!(is_trait!(super::NethunsSocketMode, Sync));
        assert!(is_trait!(super::NethunsSocketOptions, Sync));
        assert!(is_trait!(super::NethunsStat, Sync));
    }
    
    
    #[test]
    fn test_nethuns_socket_options_builder() {
        let numblocks: u32 = 12;
        let opt1 = super::NethunsSocketOptionsBuilder::default()
            .numblocks(numblocks)
            .build()
            .unwrap();
        let mut opt2 = super::NethunsSocketOptions::default();
        opt2.numblocks = numblocks;
        assert_eq!(opt1, opt2);
    }
}
