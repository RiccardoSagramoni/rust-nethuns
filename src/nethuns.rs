use std::ffi::{CStr, CString};
use std::mem;
use std::os::fd::AsRawFd;

use num_traits::AsPrimitive;
use rustix::net;

use crate::global::{NethunsNetInfo, NETHUNS_GLOBAL};


///
pub fn __nethuns_set_if_promisc(devname: &CString) -> Result<(), String> {
    let mut flags =
        nethuns_ioctl_if(devname, SocketConfigurationFlag::SIOCGIFFLAGS, 0)
            .map_err(|e| {
                format!(
                    "[__nethuns_set_if_promisc] nethuns_ioctl_if failed: {e}"
                )
            })?;
    
    if let Ok(mut mutex_guard) = NETHUNS_GLOBAL.lock() {
        let info = match mutex_guard.get_mut(devname) {
            Some(info) => info,
            None => {
                mutex_guard
                    .insert(devname.to_owned(), NethunsNetInfo::default());
                let info = mutex_guard.get_mut(devname).ok_or(format!(
                    "failed to get info for device {:?}",
                    devname
                ))?;
                info.promisc_refcnt = if (flags & libc::IFF_PROMISC as u32) == 0
                {
                    0
                } else {
                    1
                };
                info
            }
        };
        
        info.promisc_refcnt += 1;
        
        let do_promisc = (flags & libc::IFF_PROMISC as u32) == 0;
        
        if do_promisc {
            flags |= libc::IFF_PROMISC as u32;
            if let Err(e) = nethuns_ioctl_if(
                devname,
                SocketConfigurationFlag::SIOCSIFFLAGS,
                flags,
            ) {
                info.promisc_refcnt -= 1;
                return Err(format!(
                    "[__nethuns_set_if_promisc] nethuns_ioctl_if failed: {e}"
                ));
            }
        }
        
        if do_promisc {
            eprintln!("device {:?} promisc mode set", devname);
        } else {
            eprintln!("device {:?} (already) promisc mode set", devname);
        }
    };
    
    Ok(())
}


///
fn nethuns_ioctl_if(
    devname: &CStr,
    what: SocketConfigurationFlag,
    flags: u32,
) -> Result<u32, String> {
    let socket =
        net::socket(net::AddressFamily::INET, net::SocketType::DGRAM, None)
            .map_err(|e| {
                format!("[nethuns_ioctl_if] could not open socket: {e}")
            })?;
    
    let mut ifr: libc::ifreq = unsafe { mem::zeroed() };
    devname
        .to_bytes_with_nul()
        .iter()
        .take(ifr.ifr_name.len() - 1)
        .enumerate()
        .for_each(|(i, c)| {
            ifr.ifr_name[i] = *c as i8;
        });
    
    if what == SocketConfigurationFlag::SIOCSIFFLAGS {
        ifr.ifr_ifru.ifru_flags = flags as i16;
    }
    
    let ret = unsafe {
        libc::ioctl(
            socket.as_raw_fd(),
            what as u64,
            &ifr as *const libc::ifreq as *const libc::c_void,
        )
    };
    if ret < 0 {
        // FIXME nethuns_perror(nethuns_socket(s)->errbuf, "ioctl");
        return Err(format!(
            "[nethuns_ioctl_if] ioctl({:?}, {:?}, {:?}) failed",
            socket, what, ifr
        ));
    }
    
    Ok(if what == SocketConfigurationFlag::SIOCGIFFLAGS {
        unsafe { ifr.ifr_ifru.ifru_flags as u32 }
    } else {
        flags
    })
}


#[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
enum SocketConfigurationFlag {
    SIOCGIFFLAGS,
    SIOCSIFFLAGS,
}

impl AsPrimitive<u64> for SocketConfigurationFlag {
    fn as_(self) -> u64 {
        match self {
            Self::SIOCGIFFLAGS => libc::SIOCGIFFLAGS,
            Self::SIOCSIFFLAGS => libc::SIOCSIFFLAGS,
        }
    }
}
