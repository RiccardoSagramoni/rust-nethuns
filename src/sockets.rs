use core::fmt::Debug;

use crate::types::{NethunsQueue, NethunsSocketOptions};

use self::{errors::{NethunsBindError, NethunsOpenError, NethunsRecvError}, base::NethunsSocketBase, ring_slot::NethunsRingSlot};

pub mod base;
pub mod errors;
pub mod ring;
pub mod ring_slot;
pub mod types;


// Import the structs defined for the required I/O framework
cfg_if::cfg_if! {
    if #[cfg(feature="netmap")] {
        mod netmap;
        
        use netmap::NethunsSocketNetmap;
        
        pub use netmap::Pkthdr;
    }
    else {
        std::compile_error!("The support for the specified I/O framework is not available yet. Check the documentation for more information.");
    }
}


/// Trait which defines the public API for Nethuns sockets.
pub trait NethunsSocket: Debug {
    fn try_new(
        opt: NethunsSocketOptions,
    ) -> Result<Box<dyn NethunsSocket>, NethunsOpenError>
    where
        Self: Sized;
    
    
    fn bind(
        &mut self,
        dev: &str,
        queue: NethunsQueue,
    ) -> Result<(), NethunsBindError>;
    
    fn recv(&mut self) -> Result<(u64, Pkthdr, *const u8), NethunsRecvError>;
    
    
    //
    fn get_socket_base(&mut self) -> &mut NethunsSocketBase;
    fn nethuns_blocks_free(&mut self, slot: &NethunsRingSlot, blockid: u64) -> i32;
}


/// Factory to build objects which implements the trait NethunsSocket
pub struct NethunsSocketFactory();

impl NethunsSocketFactory {
    pub fn try_new_nethuns_socket(
        opt: NethunsSocketOptions,
    ) -> Result<Box<dyn NethunsSocket>, NethunsOpenError> {
        cfg_if::cfg_if! {
            if #[cfg(feature="netmap")] {
                return NethunsSocketNetmap::try_new(opt);
            }
            else {
                std::compile_error!("The support for the specified I/O framework is not available yet. Check the documentation for more information.");
            }
        }
    }
}
