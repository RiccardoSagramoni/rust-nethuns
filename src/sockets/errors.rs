use std::io;

use core::fmt::Debug;
use thiserror::Error;

/// Error type for [crate::sockets::nethuns_socket_open]
#[derive(Debug, Error)]
pub enum NethunsOpenError {
    #[error("[open] invalid options: {0}")]
    InvalidOptions(String),
    #[error("[open] an unexpected error occurred: {0}")]
    Error(String),
}

/// Error type for [crate::sockets::BindableNethunsSocket::bind]
#[derive(Debug, Error)]
pub enum NethunsBindError {
    #[error(
        "[bind] error caused by an illegal or inappropriate argument: {0}"
    )]
    IllegalArgument(String),
    #[error("[bind] error of the I/O framework: {0}")]
    FrameworkError(String),
    #[error("[bind] an unexpected error occurred: {0}")]
    Error(String),
}

/// Error type for [crate::sockets::NethunsSocket::recv]
#[derive(Debug, Error)]
pub enum NethunsRecvError {
    #[error("[recv] socket not in RX mode")]
    NotRx,
    #[error("[recv] ring in use")]
    InUse,
    #[error("[recv] no packets have been received")]
    NoPacketsAvailable,
    #[error("[recv] the received packet has been filtered out")]
    PacketFiltered,
    #[error("[recv] error of the I/O framework: {0}")]
    FrameworkError(String),
    #[error("[recv] an unexpected error occurred: {0}")]
    Error(String),
}

/// Error type for [crate::sockets::NethunsSocket::send]
#[derive(Debug, Error)]
pub enum NethunsSendError {
    #[error("[send] socket not in TX mode")]
    NotTx,
    #[error("[send] ring in use")]
    InUse,
    #[error("[send] an unexpected error occurred: {0}")]
    Error(String),
}


/// Error type for [crate::sockets::NethunsSocket::flush]
#[derive(Debug, Error)]
pub enum NethunsFlushError {
    #[error("[flush] socket not in TX mode")]
    NotTx,
    #[error("[flush] ring in use")]
    InUse,
    #[error("[flush] failed transmission: {0}")]
    FailedTransmission(String),
    #[error("[recv] error of the I/O framework: {0}")]
    FrameworkError(String),
    #[error("[flush] an unexpected error occurred: {0}")]
    Error(String),
}


/// Error type for [crate::sockets::base::pcap::NethunsSocketPcapTrait::open]
#[derive(Debug, Error)]
pub enum NethunsPcapOpenError {
    // STANDARD_PCAP_READER
    #[error("[pcap_open] could not open pcap file for writing (enable `NETHUNS_USE_BUILTIN_PCAP_READER` feature to use builtin pcap reader)")]
    WriteModeNotSupported,
    #[error("[pcap_open] error while parsing pcap file: {0}")]
    PcapError(String),
    
    // BUILTIN_PCAP_READER
    #[error("[pcap_open] magic pcap_file_header not supported ({0:02x})")]
    MagicNotSupported(u32),
    #[error("[pcap_open] error while using file: {0}")]
    FileError(#[from] io::Error),
}

impl<I> From<pcap_parser::PcapError<I>> for NethunsPcapOpenError
where
    I: Debug + Sized,
{
    fn from(e: pcap_parser::PcapError<I>) -> Self {
        NethunsPcapOpenError::PcapError(format!("{:?}", e))
    }
}


/// Error type for [crate::sockets::base::pcap::NethunsSocketPcapTrait::read]
#[derive(Debug, Error)]
pub enum NethunsPcapReadError {
    #[error("[pcap_read] head ring in use")]
    InUse,
    
    // STANDARD_PCAP_READER
    #[error("[pcap_read] error while parsing pcap file: {0}")]
    PcapError(String),
    
    // BUILTIN_PCAP_READER
    #[error("[pcap_read] error during access to file: {0}")]
    FileError(io::Error),
    #[error("[pcap_read] end of file")]
    Eof,
}

impl<I> From<pcap_parser::PcapError<I>> for NethunsPcapReadError
where
    I: Debug + Sized,
{
    fn from(e: pcap_parser::PcapError<I>) -> Self {
        match e {
            pcap_parser::PcapError::Eof => NethunsPcapReadError::Eof,
            _ => NethunsPcapReadError::PcapError(format!("{:?}", e)),
        }
    }
}

impl From<io::Error> for NethunsPcapReadError {
    fn from(e: io::Error) -> Self {
        match e.kind() {
            io::ErrorKind::UnexpectedEof => NethunsPcapReadError::Eof,
            _ => NethunsPcapReadError::FileError(e),
        }
    }
}


/// Error type for [crate::sockets::base::pcap::NethunsSocketPcapTrait::write]
#[derive(Debug, Error)]
pub enum NethunsPcapWriteError {
    // STANDARD_PCAP_READER
    #[error("[pcap_write] operation not supported")]
    NotSupported,
    
    // BUILTIN_PCAP_READER
    #[error("[pcap_write] error during access to file: {0}")]
    FileError(#[from] io::Error),
}


/// Error type for [crate::sockets::base::pcap::NethunsSocketPcapTrait::store]
#[derive(Debug, Error)]
pub enum NethunsPcapStoreError {
    // STANDARD_PCAP_READER
    #[error("[pcap_store] operation not supported")]
    NotSupported,
    
    // BUILTIN_PCAP_READER
    #[error("[pcap_store] error during access to file: {0}")]
    FileError(#[from] io::Error),
}


/// Error type for [crate::sockets::base::pcap::NethunsSocketPcapTrait::rewind]
#[derive(Debug, Error)]
pub enum NethunsPcapRewindError {
    // STANDARD_PCAP_READER
    #[error("[pcap_rewind] operation not supported")]
    NotSupported,
    
    // BUILTIN_PCAP_READER
    #[error("[pcap_rewind] error while using file: {0}")]
    FileError(#[from] io::Error),
}
