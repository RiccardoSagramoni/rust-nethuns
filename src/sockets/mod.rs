pub mod base;
mod bindings;
pub mod errors;
pub mod ring;
pub mod types;

cfg_if::cfg_if! {
    if #[cfg(feature="netmap")] {
        pub mod netmap;
    }
}
