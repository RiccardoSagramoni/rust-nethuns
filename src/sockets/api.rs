//! This module exposes the data structures required
//! to interact with the specified I/O framework.
//!
//! Every framework-specific implementation must provide:
//! - A struct which implements the [`BindableNethunsSocketTrait`] trait.
//! - A struct which implements the [`NethunsSocketTrait`] trait.
//! - A struct named [`Pkthdr`] which must implement the [`PkthdrTrait`] trait.
//!
//!
//! [`BindableNethunsSocketTrait`]: super::BindableNethunsSocketTrait
//! [`NethunsSocketTrait`]: super::NethunsSocketTrait
//! [`PkthdrTrait`]: super::PkthdrTrait

use crate::types::NethunsSocketOptions;

use super::errors::NethunsOpenError;
use super::BindableNethunsSocket;

cfg_if::cfg_if! {
    if #[cfg(feature="netmap")] {
        mod netmap;
        
        pub type Pkthdr = netmap::PkthdrNetmap;
    }
    else {
        std::compile_error!("The support for the specified I/O framework is not available yet. Check the documentation for more information.");
    }
}


/// Open a new Nethuns socket, by calling the `open` function
/// of the struct belonging to the I/O framework selected at compile time.
///
/// # Arguments
/// * `opt`: The options for the socket.
///
/// # Returns
/// * `Ok(BindableNethunsSocket)` - A new nethuns socket, in no error occurs.
/// * `Err(NethunsOpenError::InvalidOptions)` - If at least one of the options holds a invalid value.
/// * `Err(NethunsOpenError::Error)` - If an unexpected error occurs.
pub fn nethuns_socket_open(
    opt: NethunsSocketOptions,
) -> Result<BindableNethunsSocket, NethunsOpenError> {
    cfg_if::cfg_if! {
        if #[cfg(feature="netmap")] {
            netmap::BindableNethunsSocketNetmap::open(opt)
                .map(BindableNethunsSocket::new)
        }
        else {
            std::compile_error!("The support for the specified I/O framework is not available yet. Check the documentation for more information.");
        }
    }
}


#[cfg(test)]
mod test {
    use is_trait::is_trait;
    
    use crate::sockets::{
        BindableNethunsSocketTrait, NethunsSocketTrait, PkthdrTrait,
    };
    
    use super::*;
    
    #[test]
    /// Make sure that the NethunsSocket trait is implemented correctly.
    fn assert_nethuns_socket_trait() {
        cfg_if::cfg_if! {
            if #[cfg(feature="netmap")] {
                assert!(
                    is_trait!(
                        netmap::BindableNethunsSocketNetmap,
                        BindableNethunsSocketTrait
                    )
                );
                assert!(
                    is_trait!(
                        netmap::NethunsSocketNetmap,
                        NethunsSocketTrait
                    )
                );
            }
            else {
                std::compile_error!("The support for the specified I/O framework is not available yet. Check the documentation for more information.");
            }
        }
    }
    
    #[test]
    /// Make sure that the Pkthdr struct implements the PkthdrTrait trait.
    fn assert_pkthdr_trait() {
        assert!(is_trait!(Pkthdr, PkthdrTrait));
    }
}
