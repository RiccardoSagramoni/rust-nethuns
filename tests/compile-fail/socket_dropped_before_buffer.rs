use nethuns::sockets::BindableNethunsSocket;
use nethuns::types::{NethunsQueue, NethunsSocketOptions};

fn main() {
    let opt = NethunsSocketOptions::default();
    
    let socket = BindableNethunsSocket::open(opt)
        .unwrap()
        .bind("dev", NethunsQueue::Any)
        .unwrap();
    
    let packet = socket.recv().unwrap();
    let buffer = packet.buffer();
    drop(socket);
    drop(buffer);
}
