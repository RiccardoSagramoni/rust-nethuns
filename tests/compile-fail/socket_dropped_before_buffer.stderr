error[E0505]: cannot move out of `socket` because it is borrowed
  --> tests/compile-fail/socket_dropped_before_buffer.rs:14:10
   |
7  |     let socket = BindableNethunsSocket::open(opt)
   |         ------ binding `socket` declared here
...
12 |     let packet = socket.recv().unwrap();
   |                  ------ borrow of `socket` occurs here
13 |     let buffer = packet.packet();
14 |     drop(socket);
   |          ^^^^^^ move out of `socket` occurs here
15 |     drop(buffer);
   |          ------ borrow later used here
